### DISCUSSION TITLE
ClassName not merging correctly?

### DISCUSSION THREAD

**Discussion by idlework** [nativewind/nativewind#1323] [Q&A] [answered]
Presumably this is a merging issue with `className`, but I want to see if this is the case first.

In my example I give my button a `text-black` I expect my button to have a black text color but it remains white as noted in the button component with `text-white`.

``` javascript
import { PropsWithChildren } from "react";
import { Pressable, Text, View } from "react-native";

export default function HomeScreen() {
  return (
    <View className="flex-1 justify-center items-center">
      <Button className="bg-red-500" labelClassName="text-black">
        This is a button
      </Button>
    </View>
  );
}

type ButtonProps = {
  className?: string;
  labelClassName?: string;
} & PropsWithChildren;

function Button({ className, labelClassName, children }: ButtonProps) {
  return (
    <Pressable className={`p-4 bg-blue-500 ${className}`}>
      <Label className={`text-white ${labelClassName}`}>{children}</Label>
    </Pressable>
  );
}

type LabelProps = {
  className?: string;
} & PropsWithChildren;

function Label({ className, children }: LabelProps) {
  return <Text className={`text-black ${className}`}>{children}</Text>;
}
```

The `className` string for Label is as follows: `text-black text-white text-black`. In v2 this would result in an black text and in v4 this results in a white text. Here I must conclude that the last occurrence is not the dominant one, but what I thought what it should do.

My question is am I implementing v4 wrong? 
Or is there something I am overlooking with v4? 
Or is this an issue with the package?

You can find a working example of the issue [here](https://github.com/idlework/nativewind-classname-merging)
Example project is an Expo project @latest and NativeWind v4 installed via the installation guide.

What I know of Tailwind CSS is: the last occurrence of a utility class typically takes precedence and "dominates" when multiple classes of the same type are applied to an element.

And that Tailwind CSS works with the concept of cascading, where the last declared utility class will generally take precedence, much like in regular CSS. So, even though text-white is declared before the final text-black, the last text-black wins.

**Comment by marklawlor** [comment]
In CSS, order is determined by CSS specificity not class name order. V4 fixed a number of ordering issues so now the same order should be applied on both web and native.

The Tailwind team recommends using their prettier plugin which will rewrite the classNames to match specificity order.

**Reply by idlework** [reply, marked as answer]
Thank you for your quick response @marklawlor 

I made a test with tailwind and the operation is indeed the same.

Unfortunately this change of v4 is a breaking change with v2 which means that in my project a few hundred components have to be changed but so be it.

In addition, the prettier plugin has no impact on this because the utility classes are distributed across different components. A solution would be to use `tailwind-merge` and replace `text-black ${className}` with `twMerge("text-black", className)`. But `tailwind-merge` comes at a cost.

@marklawlor Is there a place in/with NativeWind that I can hook into to arrange this project broadly? Perhaps I'll create an extension/plugin for others experiencing the same problem.

**Reply by marklawlor** [reply]
The usage of tailwind-merge is the same as on the web, you apply it yourself at the component level.

**Reply by idlework** [reply]
Completely correct, that is also why I suggested this solution in my earlier post. This converstion comes across as if you don't read all of the messages in this conversation. Because the solutions you suggest either don't work or have already been mentioned.

But still thanks for responding.

If you could point me into a right direction with this question it would be well appreciated: Is there a place in/with NativeWind that I can hook into to arrange the sorting of utility classes project broadly? Perhaps I'll create an extension/plugin for others experiencing the same problem. Or an extension where you can specify the order of resolution.

**Reply by marklawlor** [reply]
This is a volunteer project and I give up my own free time to assist people. Please don't give me attitude. If you do not like an answer you can politely ask me to clarify.

I believe I answered your question but to clarify further. NativeWind tries to give a web-like experience across all platforms. The web does not offer a hook (or similar device) to change how `classNames` work, so NativeWind doesn't either. `tailwind-merge` currently has 3million weekly downloads, in which people use it by as described in their README

```
import { twJoin } from 'tailwind-merge'

function MyComponent({ forceHover, disabled, isMuted }) {
    return (
        <div
            className={twJoin(
                TYPOGRAPHY_STYLES_LABEL_SMALL,
                'grid w-max gap-2',
                forceHover ? 'bg-gray-200' : ['bg-white', !disabled && 'hover:bg-gray-200'],
                isMuted && 'text-gray-600',
            )}
        >
            {/* More code… */}
        </div>
    )
}
```

This is how we expect people to also use it with NativeWind

**Reply by marklawlor** [reply]
If you are using a className management library (`tailwind-variants`/`cva`/etc) I would also recommend checking their documentation, as many of them integrate with `tw-merge`

**Reply by idlework** [reply]
My apologies if I come across as rude. This is not my intention, probably I am a too direct in my communication. 

Earlier in my communication I suggested `tw-merge` as a solution for the situation. A clarification was not necessary, but thanks nonetheless because perhaps this can help others in the future.

Thank you for informing me that the other tools also support this. That could possibly keep the code a little cleaner.

**Comment by danstepanov** [comment]
Closing this discussion, if this is still relevant, feel free to open a new one.
